include/material/norsand.tcc

```diff
@@ -0,0 +1,589 @@
+    Mte_ = (6 * sin_friction_cs) / (3 + sin_friction_cs);
+
+    // Properties
+    properties_ = material_properties;
+
+  } catch (std::exception& except) {
+    console_->error("Material parameter not set: {}\n", except.what());
+  }
+}
+
+//! Initialise state variables
+template <unsigned Tdim>
+mpm::dense_map mpm::NorSand<Tdim>::initialise_state_variables() {
+  mpm::dense_map state_vars = {
+      // Mean stress
+      {"p", 0.},
+      // Deviatoric stress
+      {"q", 0.},
+      // Lode angle
+      {"lode_angle", 0.},
+      // J2
+      {"j2", 0.},
+      // J3
+      {"j3", 0.},
+      // M_theta
+      {"M_theta", Mtc_},
+      // Current void ratio
+      {"void_ratio", void_ratio_initial_},
+      // Void ratio image
+      {"e_image",
+       e_max_ - (e_max_ - e_min_) / log(crushing_pressure_ / p_image_initial_)},
+      // Image pressure
+      {"p_image", p_image_initial_},
+      // State variable psi
+      {"psi_image", void_ratio_initial_ -
+                        (e_max_ - (e_max_ - e_min_) / log(crushing_pressure_ /
+                                                          p_image_initial_))},
+      // p_cohesion
+      {"p_cohesion", p_cohesion_initial_},
+      // p_cohesion degradation function
+      {"zeta", 1.},
+      // Equivalent plastic deviatoric strain
+      {"epds", 0.},
+      // Plastic strain components
+      {"plastic_strain0", 0.},
+      {"plastic_strain1", 0.},
+      {"plastic_strain2", 0.},
+      {"plastic_strain3", 0.},
+      {"plastic_strain4", 0.},
+      {"plastic_strain5", 0.}};
+
+  return state_vars;
+}
+
+//! Compute elastic tensor
+template <unsigned Tdim>
+bool mpm::NorSand<Tdim>::compute_elastic_tensor() {
+  // Shear modulus
+  const double G = shear_modulus_;
+  const double a1 = bulk_modulus_ + (4.0 / 3.0) * G;
+  const double a2 = bulk_modulus_ - (2.0 / 3.0) * G;
+  // compute elastic stiffness matrix
+  // clang-format off
+  de_(0,0)=a1;    de_(0,1)=a2;    de_(0,2)=a2;    de_(0,3)=0;    de_(0,4)=0;    de_(0,5)=0;
+  de_(1,0)=a2;    de_(1,1)=a1;    de_(1,2)=a2;    de_(1,3)=0;    de_(1,4)=0;    de_(1,5)=0;
+  de_(2,0)=a2;    de_(2,1)=a2;    de_(2,2)=a1;    de_(2,3)=0;    de_(2,4)=0;    de_(2,5)=0;
+  de_(3,0)= 0;    de_(3,1)= 0;    de_(3,2)= 0;    de_(3,3)=G;    de_(3,4)=0;    de_(3,5)=0;
+  de_(4,0)= 0;    de_(4,1)= 0;    de_(4,2)= 0;    de_(4,3)=0;    de_(4,4)=G;    de_(4,5)=0;
+  de_(5,0)= 0;    de_(5,1)= 0;    de_(5,2)= 0;    de_(5,3)=0;    de_(5,4)=0;    de_(5,5)=G;
+  // clang-format on
+
+  return true;
+}
+
+//! Compute stress invariants
+template <unsigned Tdim>
+bool mpm::NorSand<Tdim>::compute_stress_invariants(const Vector6d& stress,
+                                                   mpm::dense_map* state_vars) {
+
+  // Note that in this subroutine, stress is compression positive
+
+  // Compute mean stress p
+  double mean_p = (stress(0) + stress(1) + stress(2)) / 3.;
+  if (mean_p < 1.0E-15) mean_p = 1.0E-15;
```

Comment:
I prefer: std::numeric_limits<double>::epsilon() instead of 1.0E-15

Reply:
I purposefully used this because we know it's double, and I want to get it to zero even before it reaches the machine epsilon. Let's just say, this could be a parameter we put in.

Reply:
Yes, it could cause `nan` because there are many instances where we divide by `mean_p`